For part 1:
# • How does the length of the hash correspond to the input string?
# • Are there any visible correlations between the hash and the input string?
# • What are the issues related to the cryptographic weakness of MD5?

My answers:

1. no matter the length of the input, the output hash length remains the same at 32
2. not that i can see other thanthe fact that it is only using the first few letters of the alphabet ? like a,b,c,d,e,f
3. MD5 is broken: attackers can create collisions and exploit length-extension, so it fails as an integrity check or signature.
Its speed also makes brute-force and rainbow-table attacks trivial—use SHA-2/3 or a slow KDF instead.


Part 2:
Using ther brute for attack, i put the hashes into an array and i then used the hint of 5 character lenth to to brute force and check against them

• How much time did you take in total? [18.85 seconds]
• How much time does it take to crack each string, on average? [1.29 seconds on avg]
• Is it possible to amortize (gradually write off the initial cost of) the brute
forcing attempts?

yes, instead of having do keep creating the and hash functions for all the possibilities of the words ecah time i run the function, i can just do it once and put that into a set or a list so i can just lookup those values there, i pay the cost greatly once and can then reap the rewards after

In [5]:
!python3 ex2.py

found aseas in 0.49 seconds (total 0.49s)
found cance in 1.47 seconds (total 1.96s)
found di5gv in 1.20 seconds (total 3.16s)
found dsmto in 0.25 seconds (total 3.41s)
found egunb in 0.65 seconds (total 4.06s)
found hed4e in 2.84 seconds (total 6.90s)
found lou0g in 4.26 seconds (total 11.17s)
found mlhdi in 0.91 seconds (total 12.08s)
found nized in 0.94 seconds (total 13.02s)
found ofror in 0.90 seconds (total 13.92s)
found opmen in 0.27 seconds (total 14.19s)
found owso9 in 0.20 seconds (total 14.39s)
found sso55 in 3.84 seconds (total 18.23s)
found tpoin in 0.91 seconds (total 19.14s)
found tthel in 0.11 seconds (total 19.25s)
found all i think please check
Time taken overall: 19.25 seconds
1.2832767963409424


Here is the output in the file

In [2]:
!cat ex2_hash.txt

aseas
cance
di5gv
dsmto
egunb
hed4e
lou0g
mlhdi
nized
ofror
opmen
owso9
sso55
tpoin
tthel


Part 3 : SALT

What are the observed differences between your ease of cracking the salted vs
the unsalted plaintexts?

- IT WAS FASTER AND EASER FOR THE SECOND ONE ? 


• Report the difference in time taken to crack the salted and the unsalted hash
values.

- it was 15.40 seconds overall for the salted one and for the non salted it was 19.25 seconds

• Explain any differences between salted and un-salted crack strategies.

- I pre-created my salted hashes file and loaded those values into a set at the start. Then in the brute-force loop I simply:

Generate a candidate (5-char core + salt)

MD5 it

Do a quick if digest in targets: check

Because targets is a set, each check is a single constant-time lookup instead of scanning through a list (or appending every time). I still loop over all 36⁵ candidates, but by removing that per-guess list scan, each iteration does far less extra work—which is why the second run feels so much faster.

In [6]:
!python3 ex3.py

Using salt = 'j' to crack 15 hashes…
Found: aseasj -> 5c7c8456d6ae041fd14f46230ab454b1 in 0.41s (total 0.41s) (1/15)
Found: cancej -> f539ed6f2ec65a31faa7b16035392f63 in 1.19s (total 1.59s) (2/15)
Found: di5gvj -> 0c006ada8fc0f5f131083007aa4e29e2 in 0.98s (total 2.57s) (3/15)
Found: dsmtoj -> ff15879bf55aa31793736fd6d37b3ee9 in 0.21s (total 2.78s) (4/15)
Found: egunbj -> 9a9c83022838c0ff2f896593bf66bc98 in 0.53s (total 3.31s) (5/15)
Found: hed4ej -> b8858eb9b1c4cb2c2fc2728d5a8ac84d in 2.30s (total 5.61s) (6/15)
Found: lou0gj -> 9692c6314554c514ad4aa78856dde2fb in 3.32s (total 8.93s) (7/15)
Found: mlhdij -> 1a54bf61d0c082e9e76fd4202c56c79a in 0.70s (total 9.64s) (8/15)
Found: nizedj -> 30e7a3d02f8bdf8bd8edf858d1932ba3 in 0.73s (total 10.36s) (9/15)
Found: ofrorj -> 8866b6ed857eb5da3be75bf4e3271201 in 0.72s (total 11.08s) (10/15)
Found: opmenj -> 080ee36f9ae4aa499a8ce7f8d7276431 in 0.22s (total 11.30s) (11/15)
Found: owso9j -> bebf7f4eaa40ae6789934c8071889bb2 in 0.16s (total 11.46s) (12/